A60-xds-stateful-session-affinity-weighted-clusters.md

```diff
@@ -0,0 +1,270 @@
+
+Because of the changes we made to the config selector we need a new filter
+to finalize cluster selection and to release refs to the cluster map and
+route config. This new filter is called `ClusterSelectionFilter` and it runs
+just before the Router filter. This is an internal implementation detail of
+gRPC so the xDS config is not aware of this new filter. Note:
+
+ * This is the second-last filter (retry/router filter being last).
+   Specifically it has to run after the `StatefulSessionFilter`.
+
+ * It gets the selected cluster name from `XdsClusterAttributeTypeName` and
+   cluster map from `XdsClusterMapAttribute`.
+
+ * It then takes a ref to that cluster entry from the map and releases the
+   ref to the cluster map.
+
+With this approach, `StatefulSessionFilter` is able to update the selected
+cluster between when the config selector selects the cluster and when the
+`ClusterSelectionFilter` takes a ref to the cluster.
+
+### Process Diagram
+
+The following diagram shows how the process works once this design is
+implemented:
+
+![Weighted Clusters Process](A60_graphics/final-diagram.png)
+
+ 1. An RPC comes in and config selector picks cluster-A (shown in bold).
+
+ 2. config selector sets 3 attributes to expose the cluster, current route,
+    cluster map and the entire route config. Note it also takes ref to the
+    cluster map and the route config.
+
+ 3. HTTP filters are invoked - specifically the stateful session affinity
+    filter that is of interest to us.
+
+ 4. the stateful session affinity filter sets the override-host attribute
+    and it also overwrites the cluster-attr based on the cookie contents.
+
+ 5. the new cluster selection filter is invoked.
+
+ 6. the cluster selection filter takes a ref to the new cluster - cluster-B
+    (shown in bold and italics) and releases the ref to the whole map and
+    the route config.
+
+ 7. Now the cluster-manager LB is invoked. It reads the cluster attribute -
+    which happens to be the correct cluster.
+
+ 8. After the usual LB hierarchy processing, the override-host LB policy is
+    invoked.
+
+ 9. The policy reads the override-host attribute and uses it to route the
+    RPC to that host which is now in the correct cluster.
+
+### Notes
+
+This design mainly deals with how two existing features - stateful session
+affinity and weighted clusters - should work together. As a result this feature
+does not require a new API or impact any existing APIs.
+
+### Temporary environment variable protection
+
+[A55: xDS-Based Stateful Session Affinity for Proxyless gRPC][A55] was only
+recently implemented and the implementation is still guarded by the
+`GRPC_EXPERIMENTAL_XDS_ENABLE_OVERRIDE_HOST` environment variable. We will
+continue to use the same environment variable to also guard this feature.
+This environment variable protection will be removed once the feature has
+proven stable.
+
+These design proposes some structural or declarative changes in the source code
```

Comment:
The only things that we need to guard with the env var are things that interact differently on the wire.  For this design, I think the only thing we need to guard is the behavior in the stateful session filter to add the cluster name to the cookie and to read the cluster name override from the cookie.

I think we can safely make the structural changes to the ConfigSelector and introduce the cluster selection filter without needing any env var guard, because those code-paths will be executed on every RPC on a xDS channel, so we'll have plenty of indication from our existing tests if those changes break anything.

Reply:
I am okay with that. Will make that change